Implement a CMake based build system for KLEE.

delcypher opened the pull request:
This is based off intial work by @jirislaby in #481. However it
has been substantially modified.

Notably it includes a separate build sytem to build the runtimes which
is inspired by the old build system. The reason for doing this is
because CMake is not well suited for building the runtime:

* CMake is configured to use the host compiler, not the bitcode
  compiler. These are not the same thing.

* Building the runtime using `add_custom_command()` is flawed
  because we can't automatically get transitive depencies (i.e.
  header file dependencies) unless the CMake generator is makefiles.

So for now we have a very simple build system for building the runtimes.
In the future we can replace this with something more sophisticated if
we need it.

Support for all important features of the old build system are implemented.

Another notable change is the CMake build system works much better with
LLVM installs which don't ship with testing tools. The build system
will download the sources for `FileCheck` and `not` tools if the
corresponding binaries aren't available and will build them. However
`lit` (availabe via `pip install lit`) and GTest must already be
installed.

Apart from better support testing a significant advantage of the
new CMake build system compared to the existing "Autoconf/Makefile"
build system is that it is **not** coupled to LLVM's build system
(unlike the existing build system).

Currently all tests pass when building with STP, Z3 or metaSMT.
I've only tried STP with different LLVM version (2.9 and 3.4). The other
solvers I only tested with LLVM 3.4.

I also quickly tried using Ninja as the CMake generator rather than
"UNIX Makefiles" and the build seemed to work okay :)

delcypher commented:
Because this shouldn't effect the existing build my intention is to merge this into master and iterate on it until it is ready to replace the existing building system (it's very nearly there). Once we are there we can start incrementally adding TravisCI testing. Once we're happy we can deprecate the old build system.

Personally I'd like to remove the old build system A.S.A.P.

MartinNowack commented:
@delcypher Great work, I'll have a look at it the next days.

ccadar commented:
@delcypher , thanks for the patch.  I think it would be good to iterate over it on GitHub, rather than merging it first into the mainline.  Thanks @MartinNowack for offering to review this.

delcypher commented:
All tests now pass under LLVM 2.9 and LLVM 3.4. I'm going to add TravisCI support.

ccadar commented:
BTW, please rebase everything into a smaller number of patches once you're happy with everything, so that we don't have things like "Fixed typo", "Tweak docs", etc.  How are things progressing?

delcypher commented:
@MartinNowack @ccadar Okay this now ready for review and AFAICT is a feature complete replacement for the existing build system.

It supports

* Building KLEE's runtimes
* Building doxygen documentation
* Building all klee libraries and tools
* Building and running unit tests and integration tests
* LLVM 3.4 (using clang 3.4 as bitcode compiler) and 2.9 (with llvm-gcc as bitcode compiler).
* STP, Z3 and metaSMT (@hoangmle as the metaSMT maintainer you might want to take a look here).
* Installation (@jirislaby you may want to take a look at this).
* TravisCI support (three configurations have been added to test the build)
* Docker image support.

I've tested on

* Ubuntu 14.04LTS with CMake 2.8.12 with gcc 4.8.4
* Arch Linux with CMake 3.6.2 gcc 6.2.1 and Clang 3.8.

@andreamattavelli I saw you mention OSX support. I've not tested that but in principle "It should just work" but it may require a few minor tweaks. Feel free to give it a try.

delcypher commented:
@ccadar Do you want me to rebase now? Due to the way I did some of these patches this pretty much has to be a monolithic patch. If that's what you really want I can do that.

MartinNowack commented:
@delcypher Great, please merge into one commit. I'll try to have a look into it today.

hoangmle commented:
@delcypher I notice a discrepancy between the old and new build system: in the old system all backends are disabled by default, while in the new one Z3 and STP are enabled by default. For testing metaSMT on Travis CI you need to disable both Z3 and STP, because otherwise metaSMT cannot be the default solver which will be used in 'make unittests' and 'make check'.

delcypher commented on cmake/find_metasmt.cmake:
```diff
@@ -0,0 +1,79 @@
+# FIXME: We should test linking
+list(APPEND KLEE_SOLVER_LIBRARIES ${metaSMT_LIBRARIES})
+
+# THIS IS HORRIBLE. The ${metaSMT_CXXFLAGS} variable
```
Does TravisCI use a specific version of metaSMT? I'm concerned about removing this code for older metaSMT versions.

hoangmle replied:
TravisCI always pulls the latest commit of metaSMT

delcypher replied:
Okay. I've kept the CMake code as it is for now but feel free to submit a PR to change the behaviour once this gets merged into master.

ccadar replied:
Is there any reason for not doing this now?

delcypher replied:
@ccadar Yes the metaSMT version used in our build scripts is not fixed. @hoangmle if you fix the build scripts to use a particular revision of meta-SMT then I'd happily make the change. Having a fixed version is important because it means we can have reproducible builds in TravisCI and in our Docker image.

delcypher commented:
@hoangmle 
> @delcypher I notice a discrepancy between the old and new build system: in the old system all backends are disabled by default, while in the new one Z3 and STP are enabled by default. For testing metaSMT on Travis CI you need to disable both Z3 and STP, because otherwise metaSMT cannot be the default solver which will be used in 'make unittests' and 'make check'.

Good catch.I wasn't sure what to do here. Having at least one solver enabled default means the configure step is more likely succeed. However solver choice is very important and the user should probably be forced explicitly set this. I'll fix this now and then do the rebase.

By the way there is a way to change which solver is being used when running integration tests by doing this

```
lit -vs --param klee_opts=-solver-backend=metasmt --param kleaver_opts=-solver-backend=metasmt  test/
```

There is no convenient way to change the solver when running the unittests.

delcypher commented:
@ccadar @MartinNowack Squash done. There are two additional commits that are kept separate because:

* The changes to `.dockerignore` are independent of this change.
* The commit adding support for git information imports third party code into KLEE's source tree so that deserves its own commit.

adrianherrera commented on CMakeLists.txt:
```diff
@@ -0,0 +1,692 @@
+endif()
+
+if ((NOT ${ENABLE_Z3}) AND (NOT ${ENABLE_STP}) AND (NOT ${ENABLE_METASMT}))
+  message(FATAL_ERROR "No solver was specified. At least one solver is required."
```
Hey guys, sorry to intrude on the PR!

I think this error message is slightly inaccurate. Should it list the options as `-DENABLE_SOLVER_STP=TRUE`, `-DENABLE_SOLVER_Z3=TRUE` and `-DENABLE_SOLVER_METASMT=TRUE`?

When I first tried to build I got this error message, so I did what it said at set `ENABLE_STP=TRUE`. And I still got the error message. It was only after looking in the CMake file that I found the option is `ENABLE_SOLVER_STP`.

delcypher replied:
@adrianherrera Don't apologise. The more eyes on this the better :) .

Well spotted this is a silly typo I made. I've fixed the mistake and also fixed a few other minor issues I noticed. If you have the time I'd love for you to give the updated PR a try and report any issues you find.

delcypher commented:
@jirislaby Thanks for the review. I've fixed all the issues you pointed out and squashed them into the main  commit. I've also rebased on KLEE master.

MartinNowack commented on CMakeLists.txt:
```diff
@@ -0,0 +1,703 @@
+# KLEE runtime support
+################################################################################
+# This is set here and not in `runtime` because `config.h` needs to be generated.
+
```
We should think about adding +Debug here as default.
It might be less confusing for the user as istats files contain the information of
the klee runtime as well. It should not hinder performance.

I'll update the autoconf side of this problem.

delcypher commented:
@MartinNowack 
> Still, my general question is, do we really want to support LLVM 3.4 and lower with this?

Yes. I want to completely remove the old build system as soon as possible. It doesn't look like LLVM 2.9 or 3.4 support is going away anytime soon so in order to make the build system change the CMake build system has to support all LLVM versions that we support right now.

Basically I want the change in the build system to be completely independent from us changing LLVM versions. The change in LLVM versions is going to take quite a long time but this build system change is ready to land right now.

> Up to 3.4, we have an existing and working build system. Starting with 3.5, LLVM officially supports cmake. In my opinion, we should do the same and build on top of that support - as you already do. 

That **sounds** like a good thing to do (using LLVM's `find_package(LLVM CONFIG)` support) but my experience using it tells me otherwise. About a year ago I tried using it and I had some many problems I eventually I had to abandon using LLVM that way from CMake because of packaging mistakes.

* The LLVM packages for Ubuntu/Debian were incorrectly shipped so the `LLVMConfig.cmake` file was completely broken (this was for LLVM 3.5). This is a result of Debian packagers deciding they wanted to change where files were installed and they did this without fixing the relevant CMake files. I reported the issue but I'm not sure if it was ever fixed.

* The official LLVM binary tarballs were also broken! I discovered last year there was a mistake in their packaging script that meant they shipped broken binary tarballs that were not usable from CMake.

There are other reasons to also rely on the `llvm-config` executable as the primary way we consume LLVM.

* `llvm-config` has been around for a long time and so is pretty well tested and is unlikely to go away.
* It is independent of how LLVM was built (i.e. with CMake or with the Autoconf/Makefile build system) that also happens to work across many different LLVM versions.
* It is less fragile than the CMake method because it is just a binary that will work wherever packagers decide to put it. The CMake method is much more fragile. I should know, I've contributed fixes upstream for it!

I left support in for consuming LLVM via `find_package(LLVM CONFIG)` because I'd like to experiment with it. It should work with recent LLVM versions (of course the rest of KLEE's code doesn't!) but we need to support building against LLVM 3.4 and 2.9 right now.

> Currently I see a lot of code of this patch that tries to cope with handling the lower LLVM version numbers.
I would rather like us to start with a much simpler state instead of handling old cruft and making the build system more complex than needed. And we need to rip it out again later. This would help us maintaining it in a longer run.

This is not as bad as you make it sound. Most of the complication is localised to the `find_llvm.cmake` file. There are a few bits of CMake code that also branch on the LLVM version (e.g. building the runtime  and llvm components to use when building the KLEE binary) but the existing build system already has this exact branching. The branching on LLVM version for the runtime can be cleaned up once I remove the old build system.

delcypher commented on CMakeLists.txt:
```diff
@@ -0,0 +1,692 @@
+list(APPEND KLEE_COMPONENT_CXX_DEFINES ${LLVM_DEFINITIONS})
+list(APPEND KLEE_COMPONENT_EXTRA_INCLUDE_DIRS ${LLVM_INCLUDE_DIRS})
+
+# FIXME: This doesn't look like it's being used.
```
There's another FIXME (relating to HAVE_INTTYPES_H, HAVE_MEMORY_H) that is exactly the same problem. If I'm going to make this change I ought to fix those too.

I'm avoiding doing it here because I want the build system to behave as similar to the old one as possible (i.e. the `config.h`) file is the.

I'd happily fix these once the old build system is removed

delcypher commented on CMakeLists.txt:
```diff
@@ -0,0 +1,692 @@
+# This is set here and not in `runtime` because `config.h` needs to be generated.
+
+
+set(available_klee_runtime_build_types
```
There is no `Debug` mode because with the new build system the runtime is **always** built with debug symbols (see `Makefile.config.cmake.bitcode.in`). I'm not sure why we would ever build the runtime without them (it would mean stack traces is error reports would have missing bits if the error went through the run time).

I also think `Debug` and `Release+Debug` is confusing because `Debug` means two different things here. In the former it means `-O0 -g` and in the later it means just `-g`.

MartinNowack replied:
At least in the autoconf build system:
* Release = Optimization enabled
* Debug = Debug information enabled
* Asserts = assertions enabled
Combine them to pick your choice.

So `Release+Debug` makes indeed sense and also `Release` makes sense.

Why would you like to disable `Debugging` information? Because it might hinder
some optimizations - and indeed it did this for older LLVM versions.
If I remember correctly, the 3.* branches don't have that problem anymore.

delcypher replied:
Okay. I'll add support for these other build modes.

delcypher replied:
Done.

delcypher commented on CMakeLists.txt:
```diff
@@ -0,0 +1,692 @@
+################################################################################
+option(ENABLE_KLEE_UCLIBC "Enable support for klee-uclibc" OFF)
+if (ENABLE_KLEE_UCLIBC)
+  message(STATUS "klee-uclibc support enabled")
```
We probably should warn. I actually found that our testing is kind of broken here too. If we configure with the posix runtime enabled but without klee-uclibc then we will try to run a bunch of integration tests that pass `-libc=uclibc` to KLEE and so fail. The CMake build system currently emits an author warning for this (posix runtime enabled but klee-uclibc disabled) problem when it is configured this way to remind us to fix this.

delcypher replied:
Okay I'll add a warning.

delcypher replied:
Done.

delcypher commented on CMakeLists.txt:
```diff
@@ -0,0 +1,692 @@
+################################################################################
+# Generate `CompileTimeInfo.h`
+################################################################################
+if (EXISTS "${CMAKE_SOURCE_DIR}/.git")
```
This already exists in the old build system except it cause klee to be re-built and re-linked everytime `make` is run. In the CMake build it is much smarter and it only rebuilds KLEE when `.git/HEAD` changes.

MartinNowack replied:
I know. I meant the `.git/HEAD` behavior in particular.

delcypher commented on README-CMake.md:
```diff
@@ -0,0 +1,83 @@
+* `KLEE_UCLIBC_PATH` (STRING) - Path to klee-uclibc root directory.
+
+* `KLEE_RUNTIME_BUILD_TYPE` (STRING) - Build type for KLEE's runtimes.
+   Can be `Release`, `Release+Asserts`, `Debug` or `Debug+Asserts`.
```
See my comment about not needing `Release+Debug` and `Release+Debug+Asserts`.

MartinNowack commented:
> > Still, my general question is, do we really want to support LLVM 3.4 and lower with this?
> Yes. I want to completely remove the old build system as soon as possible. It doesn't look like LLVM 2.9 or 3.4 support is going away anytime soon so in order to make the build system change the CMake build system has to support all LLVM versions that we support right now.

Why is there an urge to remove the current build system? It works fine up to LLVM 3.8 for source-based builds and up to 3.7. even with the binary packages.
Most people use the current build system. And a few of them use it also on older distributions. If they continue to use LLVM 2.9, 3.4, with removing the old build system, you force them to switch immediately without having a need for that. Or they stay with the older KLEE version and back-port changes.

I'm not against removing autoconf eventually. But I'm not in favor of forcing people to do that now for no good reason.

> That sounds like a good thing to do (using LLVM's find_package(LLVM CONFIG) support) but my experience using it tells me otherwise. About a year ago I tried using it and I had so many problems I eventually I had to abandon using LLVM that way from CMake because of packaging mistakes.
> 
>     The LLVM packages for Ubuntu/Debian were incorrectly shipped so the LLVMConfig.cmake file was completely broken (this was for LLVM 3.5). This is a result of Debian packagers deciding they wanted to change where files were installed and they did this without fixing the relevant CMake files. I reported the issue but I'm not sure if it was ever fixed.
> 
>     The official LLVM binary tarballs were also broken! I discovered last year there was a mistake in their packaging script that meant they shipped broken binary tarballs that were not usable from CMake.

But the points you mentioned are *only* related to the packaging! The source-based builds don't have that issue. If they do, let's fix them upstream, this helps much more people, instead of every LLVM-based project inventing its own workarounds to cope with the situation. The same is true for the packaging. And we can make sure current versions of LLVM are working correctly. For the older versions we have the autoconf system.

There is already so much knowledge in the LLVM's own build system. Think about support for different compilers and version, MacOS, Windows or different architectures like ARM*.
We should not replicate this effort, we should tap into this existing knowledge and use it.

> I left support in for consuming LLVM via find_package(LLVM CONFIG) because I'd like to experiment with it. It should work with recent LLVM versions (of course the rest of KLEE's code doesn't!) but we need to support building against LLVM 3.4 and 2.9 right now.

No we don't _need_ it, because we have that support already with the autoconf build system.
So why not use the `find_package(LLVM CONFIG)` feature for all LLVM versions, which support it? That would leverage the LLVM knowledge, similar like we do it currently with the `Makefile.rules` from LLVM for the autoconf build system.

Finally, just to be clear, I'm not arguing against the excellent work you are doing. It really is fantastic. But I'm trying to evaluate the necessity of some parts of it.

delcypher commented:
> Why is there an urge to remove the current build system? It works fine up to LLVM 3.8 for source-based builds and up to 3.7. even with the binary packages.

It does not work fine. There are so many hacks in the TravisCI and Dockerfile builds due to the old build system that I'm tired of maintaining. The new build system has none of these problems.

> Most people use the current build system.

You mean **all**, they have not had a choice. Up until now if you wanted to build KLEE you had to use the old build system. 

> And a few of them use it also on older distributions. If they continue to use LLVM 2.9, 3.4, with removing the old build system, you force them to switch immediately without having a need for that. Or they stay with the older KLEE version and back-port changes.

Although there are obviously issues with changing the build system. I deliberately made sure that the build works with CMake with 2.8.12 and newer so that it works fine on Ubuntu 14.04LTS.

For older distros CMake binaries are available so it will still be possible to build KLEE on older distros.

We would not be forcing them to switch immediately unless they are tracking the git master branch. To be clear this PR is not actually removing the autoconf build system. That is going to be a separate PR.

> I'm not against removing autoconf eventually. But I'm not in favor of forcing people to do that now for no good reason.

There are plenty of reasons to remove the current build system:

* It requires loads of hacks in TravisCI and the Docker image which I want to remove.
* Not very friendly for new users. I've heard lots of complaints over the years about how difficult it is to build KLEE. The new build system tries to be more user friendly.
* We have limited resources. I don't want to spend them maintaining two build systems.

Maybe it would be best not to conflate this PR with removing the existing build system. In this PR the build systems co-exist so I'd rather get this merged and then work on removing the existing build system in a separate PR.

> But the points you mentioned are only related to the packaging! The source-based builds don't have that issue. If they do, let's fix them upstream, this helps much more people, instead of every LLVM-based project inventing its own workarounds to cope with the situation. The same is true for the packaging. And we can make sure current versions of LLVM are working correctly. For the older versions we have the autoconf system.

Yes the from source builds do not have this issue but I want the build to work seamlessly regardless of the LLVM being built from source or from a package. For your reference I just tried, the official llvm-3.5 package for ubuntu 14.04LTS is **still broken**.

https://bugs.launchpad.net/ubuntu/+source/llvm/+bug/1387011

You are right that it would be better that this was fixed by the package maintainers. I have absolutely no faith that this this will be done though.

>There is already so much knowledge in the LLVM's own build system. Think about support for different compilers and version, MacOS, Windows or different architectures like ARM*.
We should not replicate this effort, we should tap into this existing knowledge and use it.

How much have you actually used LLVM's `find_package(LLVM CONFIG)`? As I said before I have tried using it several times and the experience has not been good. For better or for worse I'm very familiar with it because I contributed documentation and fixes for it to LLVM upstream and I don't consider it to be much better than using the `llvm-config` binary. Sure it adds LLVM's libraries as imported targets which is convenient and you have to type less in your CMake code to interact with it but it is **not** reliable right now. I want the experience of building KLEE to be as pain-free as possible and I will **not** sacrifice that in the name of making the CMake code slightly simpler.

> > I left support in for consuming LLVM via find_package(LLVM CONFIG) because I'd like to experiment with it. It should work with recent LLVM versions (of course the rest of KLEE's code doesn't!) but we need to support building against LLVM 3.4 and 2.9 right now.

> No we don't need it, because we have that support already with the autoconf build system.

That is incorrect. If I removed support for building against LLVM by interacting with the `llvm-config` binary then the build system becomes completely useless because none of the LLVM versions we support shipped with `LLVMConfig.cmake` files (the LLVM autoconf/Makebuild system didn't start producing these until LLVM 3.5).

The whole point of this build system is to be a drop-in replacement for what we have right now. It does a good job of that right now. What you're asking me to do makes no sense.

> So why not use the find_package(LLVM CONFIG) feature for all LLVM versions, which support it? That would leverage the LLVM knowledge, similar like we do it currently with the Makefile.rules from LLVM for the autoconf build system.

Sure I could do that but I'd rather just let user decide what to do which is why I've left an option to control this. The default is to use the `llvm-config` binary because that is the most reliable way and is in fact **the only way** that will work right now because KLEE doesn't support newer LLVM versions.

In the future we could change this but now **is the wrong time** to make this change.

delcypher commented:
Just remembered. There's another reason why it would be worth replacing the old build system with the new CMake based build system. The old build system "piggy backs" off the LLVM build system which is bad because if the LLVM build system is not available then we can't build KLEE even if all of LLVM's libraries, binaries and header files are installed.

In my experience most distros don't package the Autoconf build system files (e.g. https://apps.fedoraproject.org/packages/llvm34-devel/ ). Ubuntu/Debian appear to be an exception here. The CMake build system doesn't have this problem because it is independent of LLVM's build system and thus will work in more scenarios (i.e. building against a pre-built version of LLVM provided by distribution maintainers).

ccadar commented:
Hi all, thanks again for your work, @delcypher ! 

Reading this discussion feels like revisiting Gabriel's essays, with @MartinNowack arguing for "worse is better" (simplicity first!), while @delcypher for the MIT approach :)

There are a few issues being discussed:

1) Whether to build upon LLVM's existing cmake infrastructure.  I think this would clearly be nice, but @delcypher says the support is buggy.  I'm not familiar with this, and the points raised are vague.  @delcypher , in what way your "experience has  not been good" and the system is "not reliable right now"?   What's LLVM's status here, are they working on it? @MartinNowack , do you have any experience here?

2) Whether we should support older LLVM versions with cmake.   I appreciate @MartinNowack 's point of having a simpler cmake, but given that @delcypher put effort into backward compatibility, I'll support his point here.  The one advantage is that we could already give this new build system a try. 

3) Whether to retire the old build system.  I fully agree with @MartinNowack here that there is no reason to remove the old build system now.  It works just fine and people are used to it.  And why make them pay the price of switching when there's no reward?  If they had to switch in order to, say, get KLEE working with a newer LLVM version, then sure, otherwise it seems to be a bad idea, especially since the two systems can co-exist.  Also, there might still be issues with this new build system: just trying it on my OS X El Capitan it complains that I have an old cmake version (while the old build system works just fine).

Thanks again!
Cristian

MartinNowack commented:
I already started to reply to @delcypher but let's make it short.

> There are a few issues being discussed:
> 
> 1) Whether to build upon LLVM's existing cmake infrastructure. I think this would clearly be nice, but @delcypher says the support is buggy. I'm not familiar with this, and the points raised are vague. @delcypher , in what way your "experience has not been good" and the system is "not reliable right now"? What's LLVM's status here, are they working on it? @MartinNowack , do you have any experience here?

Just to summarize my point, cmake support for LLVM < 3.8 should not be a priority as those LLVM versions - as bad as it sounds - are already outdated. Ubuntu 16.04 has LLVM 3.8 as the default supported version (everything else is multiverse). Therefore also making the cmake build system work with `find_package(LLVM CONFIG)` for older LLVM versions (e.g. LLVM 3.5) should not be a goal. And autoconf works until LLVM 3.8. Beside that, yes, I absolutely see the future in cmake.

> 2) Whether we should support older LLVM versions with cmake. I appreciate @MartinNowack 's point of having a simpler cmake, but given that @delcypher put effort into backward compatibility, I'll support his point here. The one advantage is that we could already give this new build system a try.

I agree with that completely, we should have both systems in parallel for now.
And yes, @delcypher put a lot of effort into the new build system, so we should use it.
@delcypher I guess you agree with me, that `find_package(LLVM CONFIG)` is the way to go on the long run (assuming we don't support LLVM <3.7). It provides us many nice features which we don't have to implement/maintain and which you won't get out of llvm-config.

> 3) Whether to retire the old build system. I fully agree with @MartinNowack here that there is no reason to remove the old build system now. It works just fine and people are used to it. And why make them pay the price of switching when there's no reward? If they had to switch in order to, say, get KLEE working with a newer LLVM version, then sure, otherwise it seems to be a bad idea, especially since the two systems can co-exist. Also, there might still be issues with this new build system: just trying it on my OS X El Capitan it complains that I have an old cmake version (while the old build system works just fine).

@delcypher If it makes you a bit more happy, I'll try to maintain the autoconf stuff - if needed. So you don't have to work on that. We only have to keep the .travis support for this alive.

(The reason the Debian/LLVM-provided packages are working for KLEE is that I talked to Sylvestre he is the maintainer of the Debian and LLVM packages: https://llvm.org/bugs/show_bug.cgi?id=18802 . He was absolutely forthcoming fixing the issues. So I do believe in upstream ;) )

(And last but not least, if we ever have the opportunity, let's have bear (or another beverage) all together and I'll pay!)

delcypher commented:
@ccadar 
> 1) Whether to build upon LLVM's existing cmake infrastructure. I think this would clearly be nice, but @delcypher says the support is buggy. I'm not familiar with this, and the points raised are vague. @delcypher , in what way your "experience has not been good" and the system is "not reliable right now"? What's LLVM's status here, are they working on it? @MartinNowack , do you have any experience here?

There's nothing wrong with the `find_package(LLVM CONFIG)` upstream. It works fine. The problem is packagers breaking it. To be specific last year I tried porting Halide's CMake build system to use `find_package(LLVM CONFIG)` at the very last minute I had to abandon my work because.

* The LLVM binaries on `llvm.org` were broken.
* The Official Ubuntu/Debian packages were broken (https://bugs.launchpad.net/ubuntu/+source/llvm/+bug/1387011). They have been broken for ~ 2 years despite the issue being reported.
* The LLVM maintained Ubuntu/Debian packages were broken. I believe these have been fixed (https://llvm.org/bugs/show_bug.cgi?id=23352).

The developers of Halide didn't want users to have problems building Halide just because of packaging issues so the work was abandoned and we continued to use `llvm-config` as the way of discovering LLVM from CMake.

> 2) Whether we should support older LLVM versions with cmake. I appreciate @MartinNowack 's point of having a simpler cmake, but given that @delcypher put effort into backward compatibility, I'll support his point here. The one advantage is that we could already give this new build system a try.

:)

> 3) Whether to retire the old build system. I fully agree with @MartinNowack here that there is no reason to remove the old build system now. It works just fine and people are used to it. And why make them pay the price of switching when there's no reward? If they had to switch in order to, say, get KLEE working with a newer LLVM version, then sure, otherwise it seems to be a bad idea, especially since the two systems can co-exist. Also, there might still be issues with this new build system: just trying it on my OS X El Capitan it complains that I have an old cmake version (while the old build system works just fine).

Okay so it looks like they'll have to co-exist for some time. However I intended to update all the documentation to build via CMake so we should remove it eventually.

As for a "reward". The "reward" is that the build will succeed in scenarios where the old build system would not. For example if the `FileCheck` and `not` tools are not available the CMake build system downloads them and build them before running the tests. In the old build system the tests will just fail because the tools aren't in the user's path.

For your macOS example that isn't really representative. AFAIK cmake does not (and never has) shipped with macOS. This means you must have downloaded and installed a version of CMake on your system that is > 3 years old, you should upgrade to at least 2.8.12 and give the build system another try.

delcypher commented:
@MartinNowack 
>
> There are a few issues being discussed:

>> 1) Whether to build upon LLVM's existing cmake infrastructure. I think this would clearly be nice, but @delcypher says the support is buggy. I'm not familiar with this, and the points raised are vague. @delcypher , in what way your "experience has not been good" and the system is "not reliable right now"? What's LLVM's status here, are they working on it? @MartinNowack , do you have any experience here?

> Just to summarize my point, cmake support for LLVM < 3.8 should not be a priority as those LLVM versions - as bad as it sounds - are already outdated. Ubuntu 16.04 has LLVM 3.8 as the default supported version (everything else is multiverse). Therefore also making the cmake build system work with find_package(LLVM CONFIG) for older LLVM versions (e.g. LLVM 3.5) should not be a goal. And autoconf works until LLVM 3.8. Beside that, yes, I absolutely see the future in cmake.

Sure but until the rest of KLEE's code is brought up to date I can't test the build system with new LLVM versions. It "should just work" though.

>> 2) Whether we should support older LLVM versions with cmake. I appreciate @MartinNowack 's point of having a simpler cmake, but given that @delcypher put effort into backward compatibility, I'll support his point here. The one advantage is that we could already give this new build system a try.

> I agree with that completely, we should have both systems in parallel for now.
And yes, @delcypher put a lot of effort into the new build system, so we should use it.
@delcypher I guess you agree with me, that find_package(LLVM CONFIG) is the way to go on the long run (assuming we don't support LLVM <3.7). It provides us many nice features which we don't have to implement/maintain and which you won't get out of llvm-config.

Using `find_package(LLVM CONFIG)` is nicer to use and it would be nice to switch to it eventually. I don't want to do that until we know all packaging issues are resolved for the versions of LLVM we'd like to use it with.

Just to note `llvm-config` actually provides **more** information than `find_package(LLVM CONFIG)` does, not less. For example `llvm-config -cxxflags` shows the C++ compiler flags LLVM was built with. The `find_package(LLVM CONFIG)` does not have any variables with that information.

>> 3) Whether to retire the old build system. I fully agree with @MartinNowack here that there is no reason to remove the old build system now. It works just fine and people are used to it. And why make them pay the price of switching when there's no reward? If they had to switch in order to, say, get KLEE working with a newer LLVM version, then sure, otherwise it seems to be a bad idea, especially since the two systems can co-exist. Also, there might still be issues with this new build system: just trying it on my OS X El Capitan it complains that I have an old cmake version (while the old build system works just fine).

> @delcypher If it makes you a bit more happy, I'll try to maintain the autoconf stuff - if needed. So you don't have to work on that. We only have to keep the .travis support for this alive.

I'd like that. It would be nice to clean up the TravisCI build eventually though.

> (The reason the Debian/LLVM-provided packages are working for KLEE is that I talked to Sylvestre he is the maintainer of the Debian and LLVM packages: https://llvm.org/bugs/show_bug.cgi?id=18802 . He was absolutely forthcoming fixing the issues. So I do believe in upstream ;) )

Thanks for doing that.

> (And last but not least, if we ever have the opportunity, let's have bear (or another beverage) all together and I'll pay!)

SGTM

delcypher commented:
@ccadar @MartinNowack 
Is there anything left blocking this PR? This PR doesn't try to remove the old build system so unless there are any objections I'd like to

* Rebase and squash the commits
* Check the build still works
* Merge.

Sound good?

delcypher commented:
I've squashed and the TravisCI is rerunning.

MartinNowack commented:
I'm fine with it. Great job @delcypher !

delcypher commented:
Thanks. Merging now.

ccadar commented:
Thanks, Dan, I'm sure you're happy to see it merged ;)  The only remaining thing is to add documentation to the website.  However, please add it as a separate page, keeping the old instructions available.  The pages should be restructured so that we don't duplicate LLVM/STP installation instructions, but we can discuss this in a separate PR.  Thanks again for the patch.

ryosa commented:
Hello. I've just built this with cmake 3.6.3, and am wondering if the line below can be amended.

klee/cmake/find_llvm.cmake Line 153 
`    _run_llvm_config(_llvm_libs "--libfiles" ${ARGN})`

I deleted ${ARGN} otherwise cmake terminates by execution error of llvm-config.

delcypher commented:
@ryosa Sorry I didn't see this. Please file an issue if this is a problem. To answer your question, no I can't remove `${ARGN}` here it is **essential** because the component names have to be passed to `llvm-config`.

ryosa commented:
@delcypher I think PR #494 solved this.

ccadar commented on tools/ktest-tool/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+
+# Copy into the build directory's binary directory
+# so integration tests can find it
+configure_file(ktest-tool "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/ktest-tool" COPYONLY)
```
@delcypher , why do you need both install(...) and this line?  Shouldn't the install line copy the tool in `bin`?  This need to be better documented.  It might be useful to add some notes in the Developer's Guide about the steps needed say, to add a new tool, or use the tool in the tests.

delcypher replied:
Developers should just read the CMake documentation it's pretty clear that `install()` and `configure_file()` are different.

`install()` is for installing files when the `make install` target is invoked. The `configure_file()` line is to copy the file into the `bin/` directory in the build tree such that if the original source file (`ktest-tool` here) is modified it triggers CMake to automatically re-configure itself and copy the modified file over.
